src/OpenTelemetry.Instrumentation.EntityFrameworkCore/Implementation/EntityFrameworkDiagnosticListener.cs

```diff
@@ -0,0 +1,207 @@
+// <copyright file="EntityFrameworkDiagnosticListener.cs" company="OpenTelemetry Authors">
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// </copyright>
+
+using System;
+using System.Data;
+using System.Diagnostics;
+using OpenTelemetry.Trace;
+
+namespace OpenTelemetry.Instrumentation.EntityFrameworkCore.Implementation
+{
+    internal sealed class EntityFrameworkDiagnosticListener : ListenerHandler
+    {
+        internal const string DiagnosticSourceName = "Microsoft.EntityFrameworkCore";
+
+        internal const string ActivitySourceName = "OpenTelemetry.EntityFrameworkCore";
+        internal const string ActivityName = ActivitySourceName + ".Execute";
+
+        // TODO: get this value from payload
+        internal const string DatabaseSystemName = "mssql";
+
+        internal static readonly ActivitySource SqlClientActivitySource = new ActivitySource(ActivitySourceName, Version.ToString());
+
+        private static readonly Version Version = typeof(EntityFrameworkDiagnosticListener).Assembly.GetName().Version;
+
+        private readonly PropertyFetcher commandFetcher = new PropertyFetcher("Command");
+        private readonly PropertyFetcher connectionFetcher = new PropertyFetcher("Connection");
+        private readonly PropertyFetcher dataSourceFetcher = new PropertyFetcher("DataSource");
+        private readonly PropertyFetcher databaseFetcher = new PropertyFetcher("Database");
+        private readonly PropertyFetcher commandTypeFetcher = new PropertyFetcher("CommandType");
+        private readonly PropertyFetcher commandTextFetcher = new PropertyFetcher("CommandText");
+        private readonly PropertyFetcher exceptionFetcher = new PropertyFetcher("Exception");
+
+        private readonly EntityFrameworkInstrumentationOptions options;
+
+        public EntityFrameworkDiagnosticListener(string sourceName, EntityFrameworkInstrumentationOptions options)
```

Comment:
I see one potential issue. The SqlClient .NET code does not start Activities, it just fires events. This is why the `SqlClientDiagnosticListener` has `SupportsNullActivity => true` and is written to create Activity, not just populate them like some of the other instrumentation. This class is missing `SupportsNullActivity => true` but is still written in that style to create Activity always. Probably a mismatch somewhere. I need to go look at the EF code to figure out the correct behavior, will do later today.

Reply:
Update: I went digging through the source. EF does not spin up Activity:
https://github.com/dotnet/efcore/blob/7d6cb86ff1facb1480a51519f8052043977831bc/src/EFCore.Relational/Diagnostics/RelationalLoggerExtensions.cs#L235-L238

So the code needed `SupportsNullActivity => true`. Since I was in there, I built a unit test on top of Sqlite in-memory DB. That led to some other bug fixes. Sorry @Place1 for jumping in there, I hope you don't mind.

I just pushed what I have so far. Could probably use a bit more unit testing and maybe some enhancements to the data being collected, but working well-enough I think we can merge PR and do that separate? Or move it to contrib?